### PR TITLE
Clap 4.2

### PR DESCRIPTION
**Pull request by anfedotoff**

### CODE REVIEW

**Comment by anfedotoff** [comment]
Closes #54

File: casr/src/bin/casr-cluster.rs [hunk 1]
```diff
@@ -369,10 +359,9 @@ fn main() -> Result<()> {
             Arg::new("deduplication")
                 .short('d')
                 .long("deduplicate")
-                .takes_value(true)
-                .min_values(1)
-                .max_values(2)
-                .value_name("INPUT_DIR> <OUTPUT_DIR")
+                .action(ArgAction::Set)
+                .num_args(1..=2)
+                .value_names(["INPUT_DIR", "OUTPUT_DIR"])
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:364]: do we need to update usage.md?

File: casr/src/bin/casr-cluster.rs [hunk 2]
```diff
@@ -404,59 +392,56 @@ fn main() -> Result<()> {
...
-    let jobs = if let Some(jobs) = matches.value_of("jobs") {
-        jobs.parse::<usize>().unwrap()
+    let jobs = if let Some(jobs) = matches.get_one::<u32>("jobs") {
+        *jobs as usize
     } else {
         std::cmp::max(1, num_cpus::get() / 2)
     };
 
-    if let Some(path) = matches.value_of("ignore") {
+    if let Some(path) = matches.get_one::<String>("ignore") {
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:408]: we may parse path in option parser

File: casr/src/bin/casr-cluster.rs [hunk 3]
```diff
@@ -404,59 +392,56 @@ fn main() -> Result<()> {
...
         util::add_custom_ignored_frames(Path::new(path))?;
     }
-    if matches.is_present("similarity") {
+    if matches.contains_id("similarity") {
         let casreps: Vec<&Path> = matches
-            .values_of("similarity")
+            .get_many::<String>("similarity")
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:413]: Parse paths?

File: casr/src/bin/casr-cluster.rs [hunk 4]
```diff
@@ -404,59 +392,56 @@ fn main() -> Result<()> {
...
             .collect();
         println!(
             "{0:.5}",
             similarity(&stacktrace(casreps[0])?, &stacktrace(casreps[1])?)
         );
-    } else if matches.is_present("clustering") {
+    } else if matches.contains_id("clustering") {
         let paths: Vec<&Path> = matches
-            .values_of("clustering")
+            .get_many::<String>("clustering")
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:423]: path

File: casr/src/bin/casr-cluster.rs [hunk 5]
```diff
@@ -404,59 +392,56 @@ fn main() -> Result<()> {
...
             .map(Path::new)
             .collect();
 
         let result = make_clusters(paths[0], paths.get(1).cloned(), jobs)?;
         println!("Number of clusters: {result}");
-    } else if matches.is_present("deduplication") {
+    } else if matches.contains_id("deduplication") {
         let paths: Vec<&Path> = matches
-            .values_of("deduplication")
+            .get_many::<String>("deduplication")
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:432]: path

File: casr/src/bin/casr-cluster.rs [hunk 6]
```diff
@@ -404,59 +392,56 @@ fn main() -> Result<()> {
...
             .map(Path::new)
             .collect();
         let (before, after) = deduplication(paths[0], paths.get(1).map(|x| x.to_path_buf()), jobs)?;
         println!("Number of reports before deduplication: {before}");
         println!("Number of reports after deduplication: {after}");
-    } else if matches.is_present("merge") {
-        let paths: Vec<&Path> = matches.values_of("merge").unwrap().map(Path::new).collect();
+    } else if matches.contains_id("merge") {
+        let paths: Vec<&Path> = matches
+            .get_many::<String>("merge")
```
Comment (SweetVishnya) [inline casr/src/bin/casr-cluster.rs:441]: path

File: casr/src/bin/casr-core.rs [hunk 7]
```diff
@@ -54,76 +43,77 @@ fn main() -> Result<()> {
...
-    if mode == "offline" {
-        if !matches.is_present("output") && !matches.is_present("stdout") {
+    let mode = matches.get_one::<String>("mode").unwrap();
+    if *mode == "offline" {
+        if !matches.contains_id("output") && !matches.contains_id("stdout") {
             bail!("--stdout or --output should be specified in offline mode.");
         }
 
-        let core_path = PathBuf::from(matches.value_of("file").unwrap());
+        let core_path = PathBuf::from(matches.get_one::<String>("file").unwrap());
```
Comment (SweetVishnya) [inline casr/src/bin/casr-core.rs:116]: path

File: casr/src/bin/casr-core.rs [hunk 8]
```diff
@@ -135,8 +125,8 @@ fn main() -> Result<()> {
             .with_context(|| format!("Couldn't read core: {}", core_path.display()))?;
         let mut report = CrashReport::new();
 
-        if matches.is_present("executable") {
-            let executable_path = PathBuf::from(matches.value_of("executable").unwrap());
+        if matches.contains_id("executable") {
+            let executable_path = PathBuf::from(matches.get_one::<String>("executable").unwrap());
```
Comment (SweetVishnya) [inline casr/src/bin/casr-core.rs:129]: path

File: casr/src/bin/casr-core.rs [hunk 9]
```diff
@@ -148,16 +138,16 @@ fn main() -> Result<()> {
         let result = analyze_coredump(&mut report, &core, &core_path);
 
         if result.is_ok() {
-            if matches.is_present("output") {
-                let result_path = PathBuf::from(matches.value_of("output").unwrap());
+            if matches.contains_id("output") {
+                let result_path = PathBuf::from(matches.get_one::<String>("output").unwrap());
```
Comment (SweetVishnya) [inline casr/src/bin/casr-core.rs:142]: path

File: casr/src/bin/casr-core.rs [hunk 10]
```diff
@@ -148,16 +138,16 @@ fn main() -> Result<()> {
         let result = analyze_coredump(&mut report, &core, &core_path);
 
         if result.is_ok() {
-            if matches.is_present("output") {
-                let result_path = PathBuf::from(matches.value_of("output").unwrap());
+            if matches.contains_id("output") {
```
Comment (SweetVishnya) [inline casr/src/bin/casr-core.rs:141]: if let Some(

File: casr/src/bin/casr-core.rs [hunk 11]
```diff
@@ -54,76 +43,77 @@ fn main() -> Result<()> {
...
             .short('P')
             .hide(true)
             .long("host-pid")
             .value_name("HPID")
             .help("PID of dumped process, as seen in the initial PID namespace (since Linux 3.12)")
-            .takes_value(true))
+            .action(ArgAction::Set))
         .group(ArgGroup::new("online_analysis")
-            .args(&["core","uid","pid", "hpid"])
+            .args(["core","uid","pid", "hpid"])
```
Comment (SweetVishnya) [inline casr/src/bin/casr-core.rs:100]: don't we need parsers here?

File: casr/src/bin/casr-core.rs [hunk 12]
```diff
@@ -199,22 +189,18 @@ fn main() -> Result<()> {
...
-    let pid = matches.value_of("pid").unwrap().parse::<i32>().unwrap();
-    let culimit = matches
-        .value_of("core")
-        .unwrap_or("0")
-        .parse::<i32>()
-        .unwrap_or(-1);
-    let uid = matches.value_of("uid").unwrap().parse::<u32>().unwrap();
-    let gid = matches.value_of("gid").unwrap().parse::<u32>().unwrap();
+    let pid = *matches.get_one::<i32>("pid").unwrap();
+    let culimit = *matches.get_one::<i32>("core").unwrap_or(&-1);
```
Comment (SweetVishnya) [inline casr/src/bin/casr-core.rs:199]: there was different logic: when core is not present, the default value is 0, we may just set default value with clap

File: casr/src/bin/casr-gdb.rs [hunk 13]
```diff
@@ -40,54 +36,55 @@ fn main() -> Result<()> {
...
-        argvs.collect()
+    let argv: Vec<&str> = if let Some(argvs) = matches.get_many::<String>("ARGS") {
+        argvs.map(|s| s.as_str()).collect()
     } else {
         bail!("Wrong arguments for starting program");
     };
 
     init_ignored_frames!("cpp", "rust");
-    if let Some(path) = matches.value_of("ignore") {
+    if let Some(path) = matches.get_one::<String>("ignore") {
```
Comment (SweetVishnya) [inline casr/src/bin/casr-gdb.rs:87]: path

File: casr/src/bin/casr-gdb.rs [hunk 14]
```diff
@@ -40,54 +36,55 @@ fn main() -> Result<()> {
             Arg::new("output")
                 .short('o')
                 .long("output")
-                .takes_value(true)
-                .value_names(&["REPORT"])
+                .action(ArgAction::Set)
+                .value_name("REPORT")
```
Comment (SweetVishnya) [inline casr/src/bin/casr-gdb.rs:40]: parse paths here and there?

File: casr/src/bin/casr-libfuzzer.rs [hunk 15]
```diff
@@ -1,22 +1,20 @@
...
-
 use casr::util;
 
 use anyhow::{bail, Context, Result};
-use clap::{App, Arg};
+use clap::{
+    error::{ContextKind, ContextValue, ErrorKind},
+    Arg, ArgAction,
+};
+use log::{debug, error, info, warn};
```
Comment (SweetVishnya) [inline casr/src/bin/casr-libfuzzer.rs:8]: log::*

File: casr/src/bin/casr-libfuzzer.rs [hunk 16]
```diff
@@ -25,61 +23,59 @@ fn main() -> Result<()> {
...
-                }
-                Err(String::from("Couldn't parse jobs value"))
-        }))
+            .value_parser(clap::value_parser!(u32).range(1..)))
         .arg(
             Arg::new("input")
                 .short('i')
                 .long("input")
-                .takes_value(true)
+                .action(ArgAction::Set)
```
Comment (SweetVishnya) [inline casr/src/bin/casr-libfuzzer.rs:41]: parse path

File: casr/src/bin/casr-libfuzzer.rs [hunk 17]
```diff
@@ -25,61 +23,59 @@ fn main() -> Result<()> {
...
-                    Ok(())
+                    Ok(i_dir.to_owned())
                 })
         )
         .arg(
             Arg::new("output")
                 .short('o')
                 .long("output")
-                .takes_value(true)
+                .action(ArgAction::Set)
```
Comment (SweetVishnya) [inline casr/src/bin/casr-libfuzzer.rs:64]: parse path

File: casr/src/bin/casr-python.rs [hunk 18]
```diff
@@ -60,52 +58,53 @@ fn main() -> Result<()> {
             Arg::new("output")
                 .short('o')
                 .long("output")
-                .takes_value(true)
-                .value_names(&["REPORT"])
```
Comment (SweetVishnya) [inline casr/src/bin/casr-python.rs:64]: parse paths in different places?

File: casr/src/bin/casr-san.rs [hunk 19]
```diff
@@ -42,55 +36,56 @@ fn main() -> Result<()> {
             Arg::new("output")
                 .short('o')
                 .long("output")
-                .takes_value(true)
-                .value_names(&["REPORT"])
+                .action(ArgAction::Set)
```
Comment (SweetVishnya) [inline casr/src/bin/casr-san.rs:39]: paths?

File: casr/src/util.rs [hunk 20]
```diff
@@ -27,12 +27,12 @@ pub fn output_report(report: &CrashReport, matches: &ArgMatches, argv: &[&str])
...
     let repstr = serde_json::to_string_pretty(&report).unwrap();
 
-    if matches.is_present("stdout") {
+    if matches.get_flag("stdout") {
         println!("{repstr}\n");
     }
 
-    if matches.is_present("output") {
-        let mut report_path = PathBuf::from(matches.value_of("output").unwrap());
+    if matches.contains_id("output") {
```
Comment (SweetVishnya) [inline casr/src/util.rs:34]: paths

**Review by SweetVishnya — CHANGES_REQUESTED** [review]

File: casr/src/bin/casr-core.rs (hunk 8 above)
Comment (anfedotoff) [reply casr/src/bin/casr-core.rs:129]: Should be string, because of online mode.

File: casr/src/bin/casr-libfuzzer.rs (hunk 16 above)
Comment (anfedotoff) [reply casr/src/bin/casr-libfuzzer.rs:39]: It has already a value_parser

**Comment by SweetVishnya** [comment]
CI fails

**Review by SweetVishnya — APPROVED** [review]

**Pull request merged**